### PR TITLE
[Flink Runner] Add UseDataStreamForBatch option to Flink runner to enable batch execution on DataStream API

### PR DESCRIPTION
**Pull request by jto**
This PR is just https://github.com/apache/beam/pull/25742 updated. It also include the changes in https://github.com/apache/beam/pull/28609 which should be merged first.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by jto** [comment]
Tests are failing while publishing results which seems unrelated to those changes...

**Comment by jto** [comment]
assign set of reviewers

**Comment by jto** [comment]
Hey @tvalentyn !
This PR is another improvement on the Flink runner that never got merged.
Could you please take a look at it too :pray:

File: CHANGES.md [hunk 1]
```diff
@@ -339,6 +339,9 @@ as a workaround, a copy of "old" `CountingSource` class should be placed into a
 * The Go SDK now requires Go 1.19 to build. ([#25545](https://github.com/apache/beam/pull/25545))
 * The Go SDK now has an initial native Go implementation of a portable Beam Runner called Prism. ([#24789](https://github.com/apache/beam/pull/24789))
   * For more details and current state see https://github.com/apache/beam/tree/master/sdks/go/pkg/beam/runners/prism.
+* Add `UseDataStreamForBatch` pipeline option to the Flink runner. When it is set to true, Flink runner will run batch
```
Comment (tvalentyn) [inline CHANGES.md:342]: Please update this file; this change needs to move up.

File: runners/flink/flink_runner.gradle [hunk 2]
```diff
@@ -314,12 +316,20 @@ def createValidatesRunnerTask(Map m) {
         excludeTestsMatching 'org.apache.beam.sdk.testing.TestStreamTest.testFirstElementLate'
         // https://github.com/apache/beam/issues/20844
         excludeTestsMatching 'org.apache.beam.sdk.testing.TestStreamTest.testLateDataAccumulating'
+        if (!config.streaming) {
+          // FlinkBatchExecutionInternalTimeService does not support timer registration on timer firing.
+          excludeTestsMatching 'org.apache.beam.sdk.transforms.ParDoTest$TimestampTests.testOnTimerTimestampSkew'
+        } else {
+          // https://github.com/apache/beam/issues/25485
+          excludeTestsMatching 'org.apache.beam.sdk.transforms.GroupByKeyTest$BasicTests.testAfterProcessingTimeContinuationTriggerUsingState'
```
Comment (tvalentyn) [inline runners/flink/flink_runner.gradle:324]: is this still necessary (test still flaky)?  Does this PR make the test more flaky?

Comment (jto) [reply runners/flink/flink_runner.gradle:324]: I tried all the tests in `org.apache.beam.sdk.transforms.GroupByKeyTest$BasicTests`  and `org.apache.beam.sdk.transforms.ParDoTest$TimestampTests` 10 times for each validation (`validatesRunnerBatch`, `validatesRunnerBatchWithDataStream` and `validatesRunnerStreaming`) .


`testOnTimerTimestampSkew` never passes with `validatesRunnerBatchWithDataStream` which I think explains why this test is disabled. `GroupByKeyTest$BasicTests` on the other end always passes so I think this one does not need to be excluded.

Comment (jto) [reply runners/flink/flink_runner.gradle]: I think the reason `testOnTimerTimestampSkew` fails when using the datastream API for batch is [FLINK-18647](https://issues.apache.org/jira/browse/FLINK-18647). I don't know how to fix it though. Let me know if you have ideas.

**Comment by Abacn** [comment]
It appears this change may also caused Flink ValidatesRunner to became extremely flaky: https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_ValidatesRunner_Flink.yml?page=6&query=event%3Aschedule

Although this test suite has always been flaky (~30% fail rate), prior to Oct 27th it became extremely flaky (<5% success rate)

Some example errors:
```
testAfterProcessingTimeContinuationTriggerUsingState

java.lang.AssertionError: Second Triggered sum/Values/Values/Map/ParMultiDo(Anonymous).output: 
Expected: iterable with items [<42>] in any order
     but: no match for: <84>
```

```
testTeardownCalledAfterExceptionInStartBundleStateful

java.lang.AssertionError: Expected to have a processing method throw an exception
	at org.junit.Assert.fail(Assert.java:89)
	at org.apache.beam.sdk.transforms.ParDoLifecycleTest$ExceptionThrowingFn.after(ParDoLifecycleTest.java:460)
```

```
testTriggeredLatestSingleton

org.apache.beam.sdk.Pipeline$PipelineExecutionException: java.lang.IllegalArgumentException: PCollection with more than one element accessed as a singleton view.
	at org.apache.beam.runners.flink.TestFlinkRunner.run(TestFlinkRunner.java:75)
	at org.apache.beam.sdk.Pipeline.run(Pipeline.java:321)
	at org.apache.beam.sdk.testing.TestPipeline.run(TestPipeline.java:398)
Caused by: java.lang.IllegalArgumentException: PCollection with more than one element accessed as a singleton view.
	at org.apache.beam.sdk.values.PCollectionViews$SingletonViewFn2.apply(PCollectionViews.java:412)
	at org.apache.beam.sdk.values.PCollectionViews$SingletonViewFn2.apply(PCollectionViews.java:339)
```

**Pull request merged**